src/typechecker.erl

```diff
@@ -3391,16 +3398,108 @@ refinable(_) ->
 no_guards({clause, _, _, Guards, _}) ->
     Guards == [].
 
-%% TODO: implement proper checking of guards.
-check_guards(Env, Guards) ->
+
+fail_if_unreachable_clause(VarBinds, P) ->
+    maps:fold(fun
+                  (_K,?type(none),_Acc) ->
+                      throw({type_error, unreachable_clause, P});
+                  (_K,_V,Acc) ->
+                      Acc
+              end,
+             VarBinds,
+             VarBinds).
+
+%% Here at least one GuardSeq should be true, so we calculate the least upper bound
+check_guards(Env, Guards, P) ->
+    X = lists:map(fun(GuardSeq) ->
+                          check_guards_sequence(Env, GuardSeq)
+                  end, Guards),
+    Glb = fun(_K, Ty1, Ty2) ->
+                  normalize({type, erl_anno:new(0), union, [Ty1, Ty2]}, Env#env.tenv)
```

Comment:
Lub =

Reply:
Is this code with `Guard1A, Guard1B; Guard2A, Guard2B` correct? I think it needs test cases.

Note that a guard with alternatives on different variables `when is_integer(A); is_integer(B)` can't really be refined, (except possibly if they are considered together, i.e. `{A, B} :: {integer(), _} | {_, integer()}`).

Reply:
Yeah, bad variable naming sweat_smile

About the guards, yes, very good case, when is_integer(A); is_integer(B), my code is indeed incorrectly refining both as integers. Gonna get that fix!

Reply:
Ok, I have a question about the desired behaviour:

If we have something like
```erlang
-spec fun_different_variables(any(), any()) -> any().
fun_different_variables(A,B)
  when is_integer(A); is_integer(B) ->
    A ++ [B].
```

I assume that we want that to pass, as in that case we're just not having enough information for any refinement. But what about the case:

```erlang
-spec fun_different_variables(number() | list(), number() | list()) -> any().
fun_different_variables(A,B)
  when is_integer(A); is_integer(B) ->
    A ++ [B].
```

In that case, the guards give us no information, hence `A` inside of the clause will have type `number() | list()`, which is not promising enough and so we should fail, right?

If my assumption is correct here, then I have a ready solution for that :smiley:

Reply:
AFAIK that's what we want, yes. However, `++` doesn't crash for non-list tail. It just returns an improper list. We're not supporting improper lists (yet) though.